### PR TITLE
[FEATURE] adds HighlightAuthorViewHelper to highlight an Author ()by …

### PR DESCRIPTION
**Pull request by karliwalti**
adds HighlightAuthorViewHelper to highlight an Author.

custom html is wrapped around  author name

Example in a citestyle doc: 
```
		<publications:format.highlightAuthor author="{author}" before="<b>" after="</b>" searchterms="{filter.author}" >
			<publications:format.initials>{author.firstName} </publications:format.initials>{author.lastName}
		</publications:format.highlightAuthor>
```

TODO: (done, but will be pushed later when #27 and #28  are merged to prevent conflicts)

- Add flexform to make this an option
- update citestyles

### CODE REVIEW

**Comment by kfumy** [comment]
one question to karliwalti: beside of the viewhelper, all templates still have to be changed?

File: Classes/ViewHelpers/Format/HighlightAuthorViewHelper.php [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+     * @return array
+     */
+    protected function getAuthors(string $filter): array
+    {
+          $authors = [];
+          $authorRepository = ObjectUtility::getObjectManager()->get(AuthorRepository::class);
+          foreach (GeneralUtility::intExplode(',', $filter, true) as $identifier) {
+              $authors[] = $authorRepository->findByUid($identifier);
+          }
+    //  }
```
Comment (einpraegsam) [inline Classes/ViewHelpers/Format/HighlightAuthorViewHelper.php]: Line 82 should be removed

File: Classes/ViewHelpers/Format/HighlightAuthorViewHelper.php [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+      return $authors;
+    }
+
```
Comment (einpraegsam) [inline Classes/ViewHelpers/Format/HighlightAuthorViewHelper.php]: Also Line 85 is not allowed following PRS-2 instructions

File: Classes/ViewHelpers/Format/HighlightAuthorViewHelper.php [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    /**
+     * @return string
+     */
+    public function render(): string
+    {
+        $value = implode(PHP_EOL, GeneralUtility::trimExplode(PHP_EOL, $this->renderChildren(), true));
+        if (!empty($value))  {
+          if (!empty($this->arguments['searchterms'])) {
+            $filterauthors=$this->getAuthors($this->arguments['searchterms']);
+            $author[]= $this->arguments['author'];
```
Comment (einpraegsam) [inline Classes/ViewHelpers/Format/HighlightAuthorViewHelper.php]: Space is missing before = (PSR2)

**Comment by karliwalti** [comment]
> 
> 
> one question to karliwalti: beside of the viewhelper, all templates still have to be changed?

@kfumy  yes, as mentioned citestyles need to still be updated, but I wanted to wait until other merge requests are in that might cause conflict. Since they are now in, I will add updates in coming days. 

@einpraegsam thanks for picking up the formatting conflicts.  Happy to resolve this. I tried to maintain style of previous code and will pay extra attention to https://www.php-fig.org/psr/psr-12/

**Review by kfumy — COMMENTED** [review]
there is a mismatch between flexform definition and language definition:
flexform: flexform.pi1.template.highlightAuthors
db.localllang_db.xlf: flexform.pi1.filter.highlightAuthors

template versus filter ;-)

**Comment by karliwalti** [comment]
> 
> 
> there is a mismatch between flexform definition and language definition: flexform: flexform.pi1.template.highlightAuthors db.localllang_db.xlf: flexform.pi1.filter.highlightAuthors
> 
> template versus filter ;-)

Thank you for picking this up.

I am facing the current challenge that citestyle9 is not working on my testenv from branch develop.  Is this just mine or a common issue?

**Comment by kfumy** [comment]
no, citestyle9 is broken for me too (in the FE):
TYPO3 Exception tells me:

(1/1) TYPO3Fluid\Fluid\Core\ViewHelper\Exception

Undeclared arguments passed to ViewHelper In2code\Publications\ViewHelpers\Format\MonthNameFromNumberViewHelper: suffix. Valid arguments are: month, limit
in /app/vendor/typo3fluid/fluid/src/Core/ViewHelper/AbstractViewHelper.php line 478

**Comment by kfumy** [comment]
did not find the error yet in the citestyle.

**Comment by kfumy** [comment]
on line 208 in citestyle9.html there is a suffix=" " declaration. That causes the error.

**Comment by kfumy** [comment]
ah, it is because the pull request #38, which changes the responsibel viewhelper ist not integrated.
...

**Comment by karliwalti** [comment]
> 
> 
> ah, it is because the pull request #38, which changes the responsibel viewhelper ist not integrated. ...

I removed this now with a comment to add it back when available. 
Also added style updates for citestyle4 and 9

**Comment by kfumy** [comment]
> > ah, it is because the pull request #38, which changes the responsibel viewhelper ist not integrated. ...
> 
> I removed this now with a comment to add it back when available. Also added style updates for citestyle4 and 9

it works now. in FE there is a strange whiteroom between year and page
Bildschirmfoto 2021-11-17 um 15 20 31
es

**Comment by kfumy** [comment]
today, I worked with the publications again and there is a small point, where I can´t discover the reason of misformatting.
Sometimes in a list of publications the '.' or ';' after year is correct, sometimes there is a space, look at the screenshot. It appeares on all Citestyles, here on "Vancouver for Web".
Bildschirmfoto 2021-12-16 um 16 18 32
.

**Comment by karliwalti** [comment]
I noticed a similar thing already with the author list rendering. I do not have a good explanation for it, but the cause is following: 
When rendering is done with f:render of a section in same file, it seems that line breaks are included in the html, but this is rather random, sometimes it is rendered and sometimes not.

The only fix I have for this is to put all on one line without break. Not sure if there is a better way with Fluid.

@kfumy could you check whether this fixes the issue for you with citestyle9?

**Pull request merged**